File: Backend/astra_sales/core/permissions.py
from rest_framework import permissions
from core.models.user import RoleChoices

class IsAdminOrReadOnly(permissions.BasePermission):
    """
    Custom permission to only allow Superadmins or Admins to edit/delete master data,
    but allow authenticated users to view it.
    """
    def has_permission(self, request, view):
        # Allow read-only operations for any authenticated user
        if request.method in permissions.SAFE_METHODS:
            return request.user and request.user.is_authenticated

        # Allow write operations (POST, PUT, PATCH, DELETE) only for ADMIN/SUPERADMIN
        return request.user and request.user.is_authenticated and request.user.role in (RoleChoices.SUPERADMIN, RoleChoices.ADMIN)

class CanEditEnquiry(permissions.BasePermission):
    """
    Custom permission for object-level Enquiry editing.
    Sales Reps can only edit enquiries assigned to them.
    Other roles (SUPERADMIN, ADMIN, RFQ_TRACKER) can edit all.
    """
    def has_object_permission(self, request, view, obj):
        # Read permissions are allowed to any authenticated user (or restricted at queryset level)
        if request.method in permissions.SAFE_METHODS:
            return True

        # Write permissions
        if request.user.role in [RoleChoices.SUPERADMIN, RoleChoices.ADMIN, RoleChoices.RFQ_TRACKER]:
            return True
            
        if request.user.role == RoleChoices.SALES_REP:
            return obj.sales_rep == request.user
            
        return False

File: Frontend/frontend_admin/src/config/permissions.js
export const ROLES = {
    SUPERADMIN: 'SUPERADMIN',
    ADMIN: 'ADMIN',
    SALES_REP: 'SALES_REP',
    RFQ_TRACKER: 'RFQ_TRACKER'
};

export const PERMISSIONS = {
    // Master Data
    CAN_VIEW_MASTER_DATA: [ROLES.SUPERADMIN, ROLES.ADMIN],
    CAN_MANAGE_USERS: [ROLES.SUPERADMIN],
    
    // Enquiries / Pipeline
    CAN_EDIT_ENGG: [ROLES.RFQ_TRACKER],
    CAN_EDIT_COSTING: [ROLES.RFQ_TRACKER],
    CAN_EDIT_SALES: [ROLES.RFQ_TRACKER, ROLES.SALES_REP],
    
    // Feature Visibility
    CAN_VIEW_DASHBOARD: [ROLES.SUPERADMIN, ROLES.ADMIN, ROLES.SALES_REP, ROLES.RFQ_TRACKER],
    CAN_VIEW_REPORTS: [ROLES.SUPERADMIN, ROLES.ADMIN]
};

/**
 * Check if a user role has a specific permission
 * @param {string} userRole - The role of the current user
 * @param {Array<string>} permissionRoles - Array of roles allowed for the permission
 * @returns {boolean}
 */
export const hasPermission = (userRole, permissionRoles) => {
    if (!userRole || !permissionRoles) return false;
    return permissionRoles.includes(userRole);
};
